// @carbon/web-components/es-custom/components/tabs/tab.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import content_switcher_item_default from "../content-switcher/content-switcher-item.js";
import { TABS_TYPE } from "./defs.js";
import tabs_default from "./tabs.scss.js";
import "./tabs.js";
import { html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/tabs/tab.ts
/**
* Copyright IBM Corp. 2019, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
let CDSTab = class CDSTab extends content_switcher_item_default {
	constructor(..._args) {
		super(..._args);
		this.highlighted = false;
		this.type = TABS_TYPE.REGULAR;
	}
	/**
	* Handles `slotchange` event.
	*/
	_handleSlotChange({ target }) {
		this.tabTitle = target.assignedNodes()[0].textContent;
	}
	connectedCallback() {
		if (!this.hasAttribute("role")) this.setAttribute("role", "listitem");
		super.connectedCallback();
	}
	render() {
		const { disabled, selected, tabTitle, _handleSlotChange: handleSlotChange } = this;
		return html`
      <a
        class="${prefix}--tabs__nav-link"
        role="tab"
        aria-label="${tabTitle}"
        tabindex="${selected ? 0 : -1}"
        ?disabled="${disabled}"
        aria-selected="${selected}">
        <slot @slotchange="${handleSlotChange}"></slot>
      </a>
    `;
	}
	static {
		this.styles = tabs_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSTab.prototype, "highlighted", void 0);
__decorate([property({ reflect: true })], CDSTab.prototype, "type", void 0);
__decorate([property()], CDSTab.prototype, "tabTitle", void 0);
CDSTab = __decorate([carbonElement(`${prefix}-tab`)], CDSTab);
var tab_default = CDSTab;

//#endregion
export { tab_default as default };
