// @carbon/web-components/es-custom/components/tabs/tabs-skeleton.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import tabs_default from "./tabs.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/tabs/tabs-skeleton.ts
/**
* Copyright IBM Corp. 2019, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
let CDSTabsSkeleton = class CDSTabsSkeleton extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.contained = false;
	}
	render() {
		return html`
      <ul class="${prefix}--tabs__nav">
        <slot></slot>
      </ul>
    `;
	}
	static {
		this.styles = tabs_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSTabsSkeleton.prototype, "contained", void 0);
CDSTabsSkeleton = __decorate([carbonElement(`${prefix}-tabs-skeleton`)], CDSTabsSkeleton);
var tabs_skeleton_default = CDSTabsSkeleton;

//#endregion
export { tabs_skeleton_default as default };
