// @carbon/web-components/lib/components/tabs/defs.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

//#region src/components/tabs/defs.ts
/**
* Copyright IBM Corp. 2020, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* The keyboard action categories for dropdown.
*/
let TABS_KEYBOARD_ACTION = /* @__PURE__ */ function(TABS_KEYBOARD_ACTION) {
	/**
	* Not doing any action.
	*/
	TABS_KEYBOARD_ACTION["NONE"] = "none";
	/**
	* Keyboard action to navigate back/forward.
	*/
	TABS_KEYBOARD_ACTION["NAVIGATING"] = "navigating";
	/**
	* Keyboard action to Enter/Space.
	*/
	TABS_KEYBOARD_ACTION["ACTIVATING"] = "activating";
	/**
	* Keyboard action to navigate to first tab using home key
	*/
	TABS_KEYBOARD_ACTION["HOME"] = "home";
	/**
	* Keyboard action to navigate to last tab using end key
	*/
	TABS_KEYBOARD_ACTION["END"] = "end";
	return TABS_KEYBOARD_ACTION;
}({});
/**
* Tabs types.
*/
let TABS_TYPE = /* @__PURE__ */ function(TABS_TYPE) {
	/**
	* Regular tabs.
	*/
	TABS_TYPE["REGULAR"] = "";
	/**
	* Container type.
	*/
	TABS_TYPE["CONTAINER"] = "container";
	/**
	* Contained type.
	*/
	TABS_TYPE["CONTAINED"] = "contained";
	return TABS_TYPE;
}({});

//#endregion
exports.TABS_KEYBOARD_ACTION = TABS_KEYBOARD_ACTION;
exports.TABS_TYPE = TABS_TYPE;
