// @carbon/web-components/es/components/data-table/table-cell.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import data_table_default from "./data-table.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/data-table/table-cell.ts
/**
* Copyright IBM Corp. 2019, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* Data table cell.
*
* @element cds-table-cell
*/
let CDSTableCell = class CDSTableCell extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.overflowMenuOnHover = false;
	}
	/**
	* TODO: Uncomment when Carbon fully implements sticky header
	* Specify whether the header should be sticky.
	* Still experimental: may not work with every combination of table props
	*/
	connectedCallback() {
		if (!this.hasAttribute("role")) this.setAttribute("role", "cell");
		super.connectedCallback();
	}
	render() {
		return html`<slot></slot>`;
	}
	static {
		this.styles = data_table_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true,
	attribute: "overflow-menu-on-hover"
})], CDSTableCell.prototype, "overflowMenuOnHover", void 0);
__decorate([property({ reflect: true })], CDSTableCell.prototype, "size", void 0);
CDSTableCell = __decorate([carbonElement(`${prefix}-table-cell`)], CDSTableCell);
var table_cell_default = CDSTableCell;

//#endregion
export { table_cell_default as default };
