// @carbon/web-components/es-custom/components/tag/tag-skeleton.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import { TAG_SIZE } from "./defs.js";
import tag_default from "./tag.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";

//#region src/components/tag/tag-skeleton.ts
/**
* Copyright IBM Corp. 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
let CDSTagSkeleton = class CDSTagSkeleton extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.size = TAG_SIZE.SMALL;
	}
	render() {
		return html` <span class="${classMap({
			[`${prefix}--tag`]: true,
			[`${prefix}--skeleton`]: true,
			[`${prefix}--layout--size-${this.size}`]: this.size
		})}"></span> `;
	}
	static {
		this.styles = tag_default;
	}
};
__decorate([property({
	reflect: true,
	type: String
})], CDSTagSkeleton.prototype, "size", void 0);
CDSTagSkeleton = __decorate([carbonElement(`${prefix}-tag-skeleton`)], CDSTagSkeleton);
var tag_skeleton_default = CDSTagSkeleton;

//#endregion
export { tag_skeleton_default as default };
