// @carbon/web-components/es-custom/components/select/select-skeleton.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import select_default from "./select.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/select/select-skeleton.ts
/**
* Copyright IBM Corp. 2019, 2024
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* Skeleton of number input.
*/
let CDSSelectSkeleton = class CDSSelectSkeleton extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.hideLabel = false;
	}
	render() {
		const { hideLabel } = this;
		return html`
      ${!hideLabel && html` <span class="${prefix}--label ${prefix}--skeleton"></span> `}
      <div class="${prefix}--select ${prefix}--skeleton"></div>
    `;
	}
	static {
		this.styles = select_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true,
	attribute: "hide-label"
})], CDSSelectSkeleton.prototype, "hideLabel", void 0);
CDSSelectSkeleton = __decorate([carbonElement(`${prefix}-select-skeleton`)], CDSSelectSkeleton);
var select_skeleton_default = CDSSelectSkeleton;

//#endregion
export { select_skeleton_default as default };
