// @carbon/web-components/es/components/multi-select/multi-select-item.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import dropdown_item_default from "../dropdown/dropdown-item.js";
import "../checkbox/index.js";
import multi_select_default from "./multi-select.scss.js";
import { html } from "lit";
import { property } from "lit/decorators.js";
import { ifDefined } from "lit/directives/if-defined.js";

//#region src/components/multi-select/multi-select-item.ts
/**
* Copyright IBM Corp. 2019, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* Multi select item.
*
* @element cds-multi-select-item
*/
let CDSMultiSelectItem = class CDSMultiSelectItem extends dropdown_item_default {
	constructor(..._args) {
		super(..._args);
		this.isSelectAll = false;
		this.indeterminate = false;
		this.selectionName = "";
	}
	render() {
		const { disabled, selected, selectionName, value, indeterminate } = this;
		return html`
      <div class="${prefix}--list-box__menu-item__option">
        <cds-checkbox
          tabindex="-1"
          class="${prefix}--form-item ${prefix}--checkbox-wrapper"
          .checked=${selected}
          .indeterminate=${indeterminate}
          ?disabled=${disabled}
          name=${ifDefined(selectionName || void 0)}
          value=${value}>
          <slot></slot>
        </cds-checkbox>
      </div>
    `;
	}
	/**
	* A selector that will return multi select.
	*/
	static get selectorList() {
		return `${prefix}-multi-select`;
	}
	static {
		this.styles = multi_select_default;
	}
};
__decorate([property({ type: Boolean })], CDSMultiSelectItem.prototype, "filtered", void 0);
__decorate([property({
	type: Boolean,
	attribute: "is-select-all",
	reflect: true
})], CDSMultiSelectItem.prototype, "isSelectAll", void 0);
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSMultiSelectItem.prototype, "indeterminate", void 0);
__decorate([property({ attribute: "selection-name" })], CDSMultiSelectItem.prototype, "selectionName", void 0);
CDSMultiSelectItem = __decorate([carbonElement(`${prefix}-multi-select-item`)], CDSMultiSelectItem);
var multi_select_item_default = CDSMultiSelectItem;

//#endregion
export { multi_select_item_default as default };
