// @carbon/web-components/es/components/text-input/text-input-skeleton.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import text_input_default from "./text-input.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/text-input/text-input-skeleton.ts
/**
* Copyright IBM Corp. 2019, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* @element cds-text-input-skeleton
*
* Skeleton of number input.
*/
let CDSTextInputSkeleton = class CDSTextInputSkeleton extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.hideLabel = false;
	}
	render() {
		const { hideLabel } = this;
		return html`
      ${hideLabel ? "" : html`<span class="${prefix}--label ${prefix}--skeleton"></span>`}
      <div class="${prefix}--text-input ${prefix}--skeleton"></div>
    `;
	}
	static {
		this.styles = text_input_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true,
	attribute: "hide-label"
})], CDSTextInputSkeleton.prototype, "hideLabel", void 0);
CDSTextInputSkeleton = __decorate([carbonElement(`${prefix}-text-input-skeleton`)], CDSTextInputSkeleton);
var text_input_skeleton_default = CDSTextInputSkeleton;

//#endregion
export { text_input_skeleton_default as default };
