// @carbon/web-components/es/components/progress-indicator/progress-indicator-skeleton.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { forEach } from "../../globals/internal/collection-helpers.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import progress_indicator_default from "./progress-indicator.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/progress-indicator/progress-indicator-skeleton.ts
/**
* Copyright IBM Corp. 2019, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
let CDSProgressIndicatorSkeleton = class CDSProgressIndicatorSkeleton extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.vertical = false;
	}
	updated(changedProperties) {
		if (changedProperties.has("vertical")) forEach(this.querySelectorAll(this.constructor.selectorStep), (item) => {
			item.vertical = this.vertical;
		});
	}
	render() {
		return html`<slot></slot>`;
	}
	/**
	* A selector that will return progress steps.
	*/
	static get selectorStep() {
		return `${prefix}-progress-step-skeleton`;
	}
	static {
		this.styles = progress_indicator_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSProgressIndicatorSkeleton.prototype, "vertical", void 0);
CDSProgressIndicatorSkeleton = __decorate([carbonElement(`${prefix}-progress-indicator-skeleton`)], CDSProgressIndicatorSkeleton);
var progress_indicator_skeleton_default = CDSProgressIndicatorSkeleton;

//#endregion
export { progress_indicator_skeleton_default as default };
