// @carbon/web-components/es/components/progress-indicator/progress-step-skeleton.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import iconLoader from "../../globals/internal/icon-loader.js";
import "../skeleton-text/index.js";
import progress_indicator_default from "./progress-indicator.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";
import CircleDash16 from "@carbon/icons/es/circle-dash/16.js";

//#region src/components/progress-indicator/progress-step-skeleton.ts
/**
* Copyright IBM Corp. 2019, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
let CDSProgressStepSkeleton = class CDSProgressStepSkeleton extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.vertical = false;
	}
	render() {
		return html`
      <div
        class="${prefix}--progress-step-button ${prefix}--progress-step-button--unclickable">
        ${iconLoader(CircleDash16)}
        <p class="${prefix}--progress-label">
          <cds-skeleton-text width="40px" linecount="1"></cds-skeleton-text>
        </p>
        <span class="${prefix}--progress-line"></span>
      </div>
    `;
	}
	static {
		this.styles = progress_indicator_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSProgressStepSkeleton.prototype, "vertical", void 0);
CDSProgressStepSkeleton = __decorate([carbonElement(`${prefix}-progress-step-skeleton`)], CDSProgressStepSkeleton);
var progress_step_skeleton_default = CDSProgressStepSkeleton;

//#endregion
export { progress_step_skeleton_default as default };
