// @carbon/web-components/es/components/page-header/page-header-breadcrumb.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import page_header_default from "./page-header.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";

//#region src/components/page-header/page-header-breadcrumb.ts
/**
* Copyright IBM Corp. 2025, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* Page header Breadcrumb Bar.
* @element cds-page-header-breadcrumb
*/
let CDSPageHeaderBreadcrumb = class CDSPageHeaderBreadcrumb extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.border = true;
		this.withinGrid = false;
		this.pageActionsFlush = false;
		this.contentActionsFlush = false;
	}
	render() {
		const { withinGrid } = this;
		return html`
      <div class="${prefix}--page-header__breadcrumb-bar">
        <div class="${classMap({
			[`${prefix}--css-grid`]: !withinGrid,
			[`${prefix}--subgrid ${prefix}--subgrid--wide`]: withinGrid
		})}">
          <div
            class="${prefix}--sm:col-span-4 ${prefix}--md:col-span-8 ${prefix}--lg:col-span-16 ${prefix}--css-grid-column">
            <div class="${prefix}--page-header__breadcrumb-container">
              <div class="${prefix}--page-header__breadcrumb-wrapper">
                <slot name="icon"></slot>
                <slot></slot>
              </div>
              <div class="${prefix}--page-header__breadcrumb__actions">
                <slot name="content-actions"></slot>
                <slot name="page-actions"></slot>
              </div>
            </div>
          </div>
        </div>
      </div>
    `;
	}
	static {
		this.styles = page_header_default;
	}
};
__decorate([property({ reflect: true })], CDSPageHeaderBreadcrumb.prototype, "border", void 0);
__decorate([property({
	attribute: "within-grid",
	type: Boolean
})], CDSPageHeaderBreadcrumb.prototype, "withinGrid", void 0);
__decorate([property({
	attribute: "page-actions-flush",
	type: Boolean
})], CDSPageHeaderBreadcrumb.prototype, "pageActionsFlush", void 0);
__decorate([property({
	attribute: "content-actions-flush",
	type: Boolean
})], CDSPageHeaderBreadcrumb.prototype, "contentActionsFlush", void 0);
CDSPageHeaderBreadcrumb = __decorate([carbonElement(`${prefix}-page-header-breadcrumb`)], CDSPageHeaderBreadcrumb);
var page_header_breadcrumb_default = CDSPageHeaderBreadcrumb;

//#endregion
export { page_header_breadcrumb_default as default };
