// @carbon/web-components/es/components/textarea/textarea-skeleton.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import textarea_default from "./textarea.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/textarea/textarea-skeleton.ts
/**
* Copyright IBM Corp. 2019, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* @element cds-textarea-skeleton
*
* Skeleton of text area.
*/
let CDSTextareaSkeleton = class CDSTextareaSkeleton extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.hideLabel = false;
	}
	render() {
		const { hideLabel } = this;
		return html`
      ${hideLabel ? "" : html`<span class="${prefix}--label ${prefix}--skeleton"></span>`}
      <div class="${prefix}--skeleton ${prefix}--text-area"></div>
    `;
	}
	static {
		this.styles = textarea_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true,
	attribute: "hide-label"
})], CDSTextareaSkeleton.prototype, "hideLabel", void 0);
CDSTextareaSkeleton = __decorate([carbonElement(`${prefix}-textarea-skeleton`)], CDSTextareaSkeleton);
var textarea_skeleton_default = CDSTextareaSkeleton;

//#endregion
export { textarea_skeleton_default as default };
