// @carbon/web-components/es/components/ui-shell/header-panel.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import header_default from "./header.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/ui-shell/header-panel.ts
/**
* Copyright IBM Corp. 2023, 2024
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* Header panel
*
* @element cds-header-panel
*/
let CDSHeaderPanel = class CDSHeaderPanel extends LitElement {
	render() {
		return html`<slot></slot>`;
	}
	static {
		this.shadowRootOptions = {
			...LitElement.shadowRootOptions,
			delegatesFocus: true
		};
	}
	static {
		this.styles = header_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSHeaderPanel.prototype, "expanded", void 0);
CDSHeaderPanel = __decorate([carbonElement(`${prefix}-header-panel`)], CDSHeaderPanel);
var header_panel_default = CDSHeaderPanel;

//#endregion
export { header_panel_default as default };
