// @carbon/web-components/es/components/password-input/password-input-skeleton.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import password_input_default from "./password-input.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/password-input/password-input-skeleton.ts
/**
* Copyright IBM Corp. 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* @element cds-password-input-skeleton
*
* Skeleton of password input.
*/
let CDSPasswordInputSkeleton = class CDSPasswordInputSkeleton extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.hideLabel = false;
	}
	render() {
		const { hideLabel } = this;
		return html`
      ${!hideLabel && html` <span class="${prefix}--label ${prefix}--skeleton"></span> `}
      <div class="${prefix}--text-input ${prefix}--skeleton"></div>
    `;
	}
	static {
		this.styles = password_input_default;
	}
};
__decorate([property({
	type: Boolean,
	reflect: true,
	attribute: "hide-label"
})], CDSPasswordInputSkeleton.prototype, "hideLabel", void 0);
CDSPasswordInputSkeleton = __decorate([carbonElement(`${prefix}-password-input-skeleton`)], CDSPasswordInputSkeleton);
var password_input_skeleton_default = CDSPasswordInputSkeleton;

//#endregion
export { password_input_skeleton_default as default };
