// @carbon/web-components/es-custom/components/popover/popover-content.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import { POPOVER_BACKGROUND_TOKEN } from "./defs.js";
import popover_default from "./popover.scss.js";
import { LitElement, html } from "lit";
import { property } from "lit/decorators.js";

//#region src/components/popover/popover-content.ts
/**
* Copyright IBM Corp. 2019, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* Popover.
*
* @element cds-custom-popover-content
*/
let CDSPopoverContent = class CDSPopoverContent extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.align = "";
		this.autoalign = false;
		this.dropShadow = true;
		this.border = false;
		this.highContrast = false;
		this.open = false;
		this.tabTip = false;
		this.backgroundToken = POPOVER_BACKGROUND_TOKEN.LAYER;
		this.slot = "content";
	}
	render() {
		return html`
      <span class="${prefix}--popover">
        <span class="${prefix}--popover-content" part="content" tabindex="-1">
          <slot> </slot>
          ${this.autoalign ? html`<span
                class="${prefix}--popover-caret ${prefix}--popover--auto-align"></span>` : null}
        </span>
        ${!this.autoalign ? html`<span class="${prefix}--popover-caret"></span>` : null}
      </span>
    `;
	}
	static {
		this.styles = popover_default;
	}
};
__decorate([property({
	reflect: true,
	type: String
})], CDSPopoverContent.prototype, "align", void 0);
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSPopoverContent.prototype, "autoalign", void 0);
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSPopoverContent.prototype, "caret", void 0);
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSPopoverContent.prototype, "dropShadow", void 0);
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSPopoverContent.prototype, "border", void 0);
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSPopoverContent.prototype, "highContrast", void 0);
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSPopoverContent.prototype, "open", void 0);
__decorate([property({
	type: Boolean,
	reflect: true
})], CDSPopoverContent.prototype, "tabTip", void 0);
__decorate([property({
	reflect: true,
	type: String
})], CDSPopoverContent.prototype, "backgroundToken", void 0);
__decorate([property({ reflect: true })], CDSPopoverContent.prototype, "slot", void 0);
CDSPopoverContent = __decorate([carbonElement(`${prefix}-popover-content`)], CDSPopoverContent);
var popover_content_default = CDSPopoverContent;

//#endregion
export { popover_content_default as default };
