// @carbon/web-components/es-custom/components/modal/modal-body.js
/**
 * Copyright IBM Corp. 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { prefix } from "../../globals/settings.js";
import { carbonElement } from "../../globals/decorators/carbon-element.js";
import { __decorate } from "../../_virtual/_@oxc-project_runtime@0.112.0/helpers/decorate.js";
import modal_default from "./modal.scss.js";
import { LitElement, html } from "lit";

//#region src/components/modal/modal-body.ts
/**
* Copyright IBM Corp. 2019, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* Modal body.
*
* @element cds-custom-modal-body
*/
let CDSModalBody = class CDSModalBody extends LitElement {
	constructor(..._args) {
		super(..._args);
		this.userDefinedTabindex = null;
	}
	connectedCallback() {
		super.connectedCallback?.();
		if (this.hasAttribute("tabindex")) this.userDefinedTabindex = this.getAttribute("tabindex");
		this._resizeObserver = new ResizeObserver(() => this.checkScroll());
		this._resizeObserver.observe(this);
	}
	disconnectedCallback() {
		this._resizeObserver?.disconnect();
		super.disconnectedCallback?.();
	}
	checkScroll() {
		const hasScroll = this.scrollHeight > this.clientHeight;
		const hasAutoAlign = this.querySelector("[autoalign]") !== null;
		if (this.clientHeight <= 300 || hasAutoAlign) this.setAttribute("no-fade", "");
		else this.removeAttribute("no-fade");
		if (this.userDefinedTabindex !== null) return;
		if (hasScroll) {
			this.setAttribute("tabindex", "0");
			this.setAttribute("is-scrollable", "");
		} else {
			this.removeAttribute("tabindex");
			this.removeAttribute("is-scrollable");
		}
	}
	render() {
		return html` <slot @slotchange=${this.checkScroll}></slot> `;
	}
	static {
		this.styles = modal_default;
	}
};
CDSModalBody = __decorate([carbonElement(`${prefix}-modal-body`)], CDSModalBody);
var modal_body_default = CDSModalBody;

//#endregion
export { modal_body_default as default };
